odl/discr/partition.py

```diff
@@ -766,6 +767,132 @@ def uniform_partition_fromgrid(grid, begin=None, end=None):
+
+    >>> part = uniform_partition_options(begin=0, end=1, num_nodes=4)
+    >>> part.cell_boundary_vecs
+    (array([ 0.  ,  0.25,  0.5 ,  0.75,  1.  ]),)
+    >>> part = uniform_partition_options(begin=0, num_nodes=4, cell_sides=0.5)
+    >>> part.cell_boundary_vecs
+    (array([ 0. ,  0.5,  1. ,  1.5,  2. ]),)
+    >>> part = uniform_partition_options(end=1, num_nodes=4, cell_sides=0.5)
+    >>> part.cell_boundary_vecs
+    (array([-1. , -0.5,  0. ,  0.5,  1. ]),)
+    >>> part = uniform_partition_options(begin=0, end=1, cell_sides=0.5)
+    >>> part.cell_boundary_vecs
+    (array([ 0. ,  0.5,  1. ]),)
+
+    In higher dimensions, the parameters can be given differently in
+    each axis. Where None is given, the value will be computed:
+
+    >>> part = uniform_partition_options(begin=[0, 0], end=[1, 2],
+    ...                                  num_nodes=[4, 2])
+    >>> part.cell_boundary_vecs
+    (array([ 0.  ,  0.25,  0.5 ,  0.75,  1.  ]), array([ 0.,  1.,  2.]))
+    >>> part = uniform_partition_options(begin=[0, 0], end=[1, 2],
+    ...                                  num_nodes=[None, 2],
+    ...                                  cell_sides=[0.5, None])
+    >>> part.cell_boundary_vecs
+    (array([ 0. ,  0.5,  1. ]), array([ 0.,  1.,  2.]))
+    >>> part = uniform_partition_options(begin=[0, None], end=[None, 2],
+    ...                                  num_nodes=[2, 2],
+    ...                                  cell_sides=[0.5, 0.25])
+    >>> part.cell_boundary_vecs
+    (array([ 0. ,  0.5,  1. ]), array([ 1.5 ,  1.75,  2.  ]))
+    """
+    # np.size(None) = 1
+    sizes = [np.size(p) for p in (begin, end, num_nodes, cell_sides)]
+    max_size = int(np.max(sizes))
+
+    if max_size == 1:
+        begin = [begin]
+        end = [end]
+        num_nodes = [num_nodes]
+        cell_sides = [cell_sides]
+    else:
+        if begin is None:
+            begin = [None] * max_size
+        if end is None:
+            end = [None] * max_size
+        if num_nodes is None:
+            num_nodes = [None] * max_size
+        if cell_sides is None:
+            cell_sides = [None] * max_size
+
+        begin = list(begin)
+        end = list(end)
+        num_nodes = list(num_nodes)
+        cell_sides = list(cell_sides)
+
+    for i, (b, e, n, s) in enumerate(zip(begin, end, num_nodes, cell_sides)):
+        num_params = sum(p is not None for p in (b, e, n, s))
+        if num_params < 3:
+            raise ValueError('in axis {}: expected at least 3 of the '
+                             'parameters `begin`, `end`, `num_nodes`, '
+                             '`cell_sides`, got {}'
+                             ''.format(i, num_params))
+        if b is None:
+            begin[i] = e - n * s
+        elif e is None:
+            end[i] = b + n * s
+        elif n is None:
+            n_calc = (e - b) / s
+            n_round = int(round(n_calc))
+            if not np.isclose(n_calc, n_round, atol=1e-5):
```

Comment:
Yeah, this version is "historically grown". First I had your version, then I thought "Ah, I'll just use `np.isclose` and whatever it defaults to", then I thought the default is too strict and changed `atol`, an voilà, here's the bloated version of the above code.